### ISSUE TITLE
Cannot derive Debug for struct named builder

### ISSUE THREAD

**Issue by alanhdu** [rust-lang/rust#42453]
If I try to compile:

```rust
#[derive(Debug)]
struct builder;
```

I get a very cryptic error message:

```
error[E0530]: let bindings cannot shadow unit structs

error: aborting due to previous error
```

This is with both Rust 1.17 and the current nightly.

[Playpen link](https://play.rust-lang.org/?code=%23%5Bderive(Debug)%5D%0Astruct%20builder%3B%0A%0Afn%20main()%20%7B%7D&version=stable&backtrace=0).

(Originally discovered via https://github.com/diesel-rs/diesel/issues/934)

**Comment by nagisa** [comment]
This seems like a general hygiene issue, but am not sure if the built-in `derive`s are using the new procedural macros. If they aren’t, porting may suffice. If they are, there’s a bug in our procedural macros.

cc @jseyfried?

**Comment by alanhdu** [comment]
Closed by #48934
